python/pyspark/worker.py

```diff
@@ -446,7 +446,12 @@ def process():
         pickleSer._write_with_length((aid, accum._value), outfile)
 
     # check end of stream
-    if read_int(infile) == SpecialLengths.END_OF_STREAM:
+    res = read_int(infile)
+    if sys.version >= '3' and res == SpecialLengths.END_OF_DATA_SECTION:
```

Comment:
This sounds weird. Why in Python3, there is an additional `SpecialLengths.END_OF_DATA_SECTION`?

Reply:
same question

Reply:
It is not an 'additional' `SpecialLengths.END_OF_DATA_SECTION`, both Python2\Python3 get this -1 value which generated from `PythonRunner.writeIteratorToStream`.
The bug caused only in Python3 because Python2 handled this while Python3 not, as I described in JIRA description:https://issues.apache.org/jira/browse/SPARK-26549.

Reply:
It's handled in `FramedSerializer` in Python2:
![image](https://user-images.githubusercontent.com/4833765/50757250-b0fa4a00-1299-11e9-9c3e-b8a258c2e721.png)

Reply:
Sorry for the mess @viirya, it's nothing to do with specific python version, just bug for `parrllelize(xrange(x))`. `xrange` and `range` in parallelize has different code path, in the scenario of using range, `FramedSerializer` will handle END_OF_DATA_SECTION, while using xrange, will not enter FramedSerializer and range is xrange in Python3...

Reply:
Does that mean the check of sys.version here should be removed? like it would be safer and better to handle it for Python 2 as well?

Reply:
Thanks Sean for your reply.
```
Does that mean the check of sys.version here should be removed?
```
Yes, please wait just one moment, I'm pushing the latest code and changing the PR description. The last commit is a much cleaner approach.

Reply:
Done in ab451e5. @srowen @viirya @cloud-fan